### PR TITLE
Bugfix. EMPTY_BLOCK_STRUCTURE_ERROR shouldn't trigger on empty lambdas

### LINKED ISSUE #796: EMPTY_BLOCK_STRUCTURE_ERROR shouldn't trigger on empty lambdas
## Describe the bug
Sometimes empty lambda can be used as a function that does nothing:
```kotlin
fun foo(bar: () -> Unit = {})
```
or empty function in implementation of functional interface:
```kotlin
A({})  // instead of A(Disposable {}), where Disposable is functional interface from Java
```
Diktat shouldn't issue a warning in these cases, because they are valid and can't be done differentlty. Also, kotlin copiler warns that `Unit` is redundant in `{ Unit }`.

## Environment information
* diktat version: 0.4.2

### PR DESCRIPTION
**Pull request by aktsay6**
### What's done:
  * Fixed bug

This pull request closes #796

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt [hunk 1]
```diff
@@ -28,6 +34,9 @@ class EmptyBlock(configRules: List<RulesConfig>) : DiktatRule(
         val configuration = EmptyBlockStyleConfiguration(
             configRules.getRuleConfig(EMPTY_BLOCK_STRUCTURE_ERROR)?.configuration ?: emptyMap()
         )
+        if (node.elementType == FILE) {
+            println(node.prettyPrint())
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt:38]: 😠

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt [hunk 2]
```diff
@@ -82,18 +91,33 @@ class EmptyBlock(configRules: List<RulesConfig>) : DiktatRule(
...
+    @Suppress("UnsafeCallOnNullableType")
+    private fun isLambdaUsedAsFunction(node: ASTNode): Boolean {
+        val parents = node.parents()
+        if (parents.any { it.elementType == CALL_EXPRESSION }) {
+            val callExpression = parents.find { it.elementType == CALL_EXPRESSION }!!
+            // excepting cases like list.map { }
+            return callExpression.treeParent.elementType != DOT_QUALIFIED_EXPRESSION
+        } else if (parents.any { it.elementType == LAMBDA_EXPRESSION }) {
+            val lambdaExpression = parents.find { it.elementType == LAMBDA_EXPRESSION }!!
+            // cases like A({}). Here Lambda expression is used as a value parameter.
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt:115]: I remember that we have e.g. this case: https://github.com/cqfn/diKTat/blob/d9b6b6eff3f189aca9ac15f6ba25d230f047ff0c/diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/KotlinParser.kt#L53, where the compiler suggests to remove explicit class name as redundant, but diktat would've raised an error. Please try to remove and see if you PR fixes it.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt [hunk 3]
```diff
@@ -82,18 +91,33 @@ class EmptyBlock(configRules: List<RulesConfig>) : DiktatRule(
...
                 false
             }
 
+    @Suppress("UnsafeCallOnNullableType")
+    private fun isLambdaUsedAsFunction(node: ASTNode): Boolean {
+        val parents = node.parents()
+        if (parents.any { it.elementType == CALL_EXPRESSION }) {
+            val callExpression = parents.find { it.elementType == CALL_EXPRESSION }!!
+            // excepting cases like list.map { }
+            return callExpression.treeParent.elementType != DOT_QUALIFIED_EXPRESSION
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt:112]: Please move `return` outside of `if` and add the default `return` as `else` branch

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt [hunk 4]
```diff
@@ -41,7 +50,7 @@ class EmptyBlock(configRules: List<RulesConfig>) : DiktatRule(
 
     @Suppress("UnsafeCallOnNullableType", "TOO_LONG_FUNCTION")
     private fun checkEmptyBlock(node: ASTNode, configuration: EmptyBlockStyleConfiguration) {
-        if (node.treeParent.isOverridden() || isAnonymousSamClass(node)) {
+        if (node.treeParent.isOverridden() || isAnonymousSamClass(node) || isLambdaUsedAsFunction(node)) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt:53]: Please check the description of this rule in code style, I think we should add a note about lambdas in there too.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt (hunk 4 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt:50]: Added

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt (hunk 3 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt]: Fixed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt (hunk 1 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt]: Removed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt (hunk 2 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/EmptyBlock.kt]: Fixed

**Review by petertrr — APPROVED** [review]

**Pull request merged**